### PR TITLE
(feat): add customizable org-roam-title-to-slug-fn

### PR DESCRIPTION
**Pull request by zackteo**
###### Motivation for this change
Choose the length of the title. #781

### CODE REVIEW

File: org-roam-capture.el [hunk 1]
```diff
@@ -349,7 +349,7 @@ This uses the templates defined at `org-roam-capture-templates'."
          (title (or (plist-get res :title) title-with-keys))
          (file-path (plist-get res :path)))
     (let ((org-roam-capture--info (list (cons 'title title)
-                                        (cons 'slug (org-roam--title-to-slug title))
+                                        (cons 'slug (org-roam-title-to-slug-fn title))
```
Comment (jethrokuan) [inline org-roam-capture.el:352]:
```suggestion
                                        (cons 'slug (funcall org-roam-title-to-slug-fn title))
```

Comment (zaeph) [reply org-roam-capture.el:352]:
```suggestion
                                        (cons 'slug (funcall #'org-roam-title-to-slug-fn title))
```

File: org-roam-capture.el [hunk 2]
```diff
@@ -44,7 +44,7 @@
 (declare-function  org-roam--file-path-from-id          "org-roam")
 (declare-function  org-roam--find-file                  "org-roam")
 (declare-function  org-roam--format-link                "org-roam")
-(declare-function  org-roam--title-to-slug              "org-roam")
+(declare-function  org-roam-title-to-slug-fn            "org-roam")
```
Comment (zackteo) [inline org-roam-capture.el:47]: Is this causing the linting error?

File: org-roam.el [hunk 3]
```diff
@@ -1296,7 +1304,7 @@ which takes as its argument an alist of path-completions.  See
     (if file-path
         (org-roam--find-file file-path)
       (let ((org-roam-capture--info `((title . ,title-with-tags)
-                                      (slug  . ,(org-roam--title-to-slug title-with-tags))))
+                                      (slug  . ,(funcall org-roam-title-to-slug-fn title-with-tags))))
```
Comment (jethrokuan) [inline org-roam.el:1307]:
```suggestion
                                      (slug  . ,(funcall #'org-roam-title-to-slug-fn title-with-tags))))
```

File: org-roam.el [hunk 4]
```diff
@@ -1371,7 +1379,7 @@ If DESCRIPTION is provided, use this as the link label.  See
       (when (org-roam-capture--in-process-p)
         (user-error "Nested Org-roam capture processes not supported"))
       (let ((org-roam-capture--info `((title . ,title-with-tags)
-                                      (slug . ,(org-roam--title-to-slug title-with-tags))))
+                                      (slug . ,(funcall org-roam-title-to-slug-fn title-with-tags))))
```
Comment (jethrokuan) [inline org-roam.el:1382]:
```suggestion
                                      (slug . ,(funcall #'org-roam-title-to-slug-fn title-with-tags))))
```

**Comment by zackteo** [comment]
@jethrokuan I have made the changes you asked for in the recent commit. 
Except for `org-roam-title-to-slug-function` to `#'org-roam-title-to-slug-function` where doing so will only result in "Symbol's function definition is void: org-roam-title-to-slug-function". Unless I missed something out, this should be correct

File: org-roam.el [hunk 5]
```diff
@@ -208,6 +207,14 @@ extraction methods:
               (const :tag "sub-directories" all-directories)
               (const :tag "parent directory" last-directory)))
 
+(defcustom org-roam-title-to-slug-function #'org-roam--title-to-slug
+  "Function to be used in converting a title to the filename slug.
+Function should return a filename string based on title.
+Otherwise, function will remain as default of
+`org-roam--title-to-slug'."
```
Comment (jethrokuan) [inline org-roam.el:214]:
```suggestion
Function should return a filename string based on title."
```

**Comment by jethrokuan** [comment]
LGTM otherwise!

**Comment by xeruf** [comment]
Seems this feature was lost in v2?

**Comment by kijowski** [comment]
I've probably followed similar path to you @xeruf so maybe my findings would benefit you or someone else in the future :) 
`org-roam-node-slug` seems unused, because it's not being called directly - rather it's being called as a part of template filling mechanism described [here](https://www.orgroam.com/manual.html#Org_002droam-Template-Expansion).

So to customize slug generation you have two options:
1. Overwrite `org-roam-node-slug` method with your own definition
2. Write your own `org-roam-node-X` method and use `${X}` in template definition. In my config I have following function which does what I want:
```
(cl-defmethod org-roam-node-dash-slug ((node org-roam-node))
  "Return the slug of NODE with dashes"
  (replace-regexp-in-string "_" "-" (org-roam-node-slug node)))
```

**Comment by xeruf** [comment]
ah, I solved it with advice, with the same goal as you ;)

```sh
    (defun xf/dashify-slug (slug)
      (s-replace "_" "-" slug))
    (advice-add 'org-roam-node-slug :filter-return #'xf/dashify-slug)
```

**Pull request merged**